src/node/miner.h

```diff
@@ -225,6 +231,49 @@ class BlockAssembler
     void SortForBlock(const CTxMemPool::setEntries& package, std::vector<CTxMemPool::txiter>& sortedEntries);
 };
 
+/**
+ * BlockTemplateCache provides a thread-safe interface for creating and reusing
+ * block templates with configurable cache size.
+ *
+ * The cache stores templates with their respective config options.
+ * When a block template is requested:
+ * - If a template with matching options exists and the interval
+ *   has not elapsed, the cached template is returned.
+ * - If no template exists or the interval has elapsed, a new template is generated,
+ *   stored in the cache, and returned.
+ * - After insertion, we evict the oldest template if the cache overflows.
+ *
+ * The cache inherits from CValidationInterface to receive notifications
+ * about connected blocks, which triggers cache invalidation,
+ * ensuring stale templates are not returned.
+ */
+class BlockTemplateCache : public CValidationInterface
+{
+public:
+    virtual ~BlockTemplateCache() = default;
+
+    /**
+     * If a cached template exists with identical options and its age does not
+     * exceed the specified interval, the cached template is returned.
+     * Otherwise, a new template is created and stored in the cache.
+     *
+     * @param options The block assembly options to use.
+     * @param interval Maximum age of a cached template to be considered fresh.
+     * @return std::shared_ptr<CBlockTemplate> Shared pointer to the block template.
+     */
+    virtual std::pair<std::shared_ptr<CBlockTemplate>, NodeClock::time_point> GetBlockTemplate(
+        const BlockAssembler::Options& options,
+        std::chrono::seconds interval
```

Comment:
In commit "node: add block template cache to miner" (4de7887ca47197654f480471db04750d44e405de)

Would it make sense for this to use milliseconds or another unit, instead of seconds so callers could have more granularity.

Also as suggested elsewhere I think it could make sense for interval to be a `BlockCreateOptions` field (called `max_age` or similar) so existing interfaces would not need to be modified to start taking advantage of template caching.

Reply:
This is fixed now